### PR TITLE
Search and Filter for Templates

### PR DESCRIPTION
**Pull request by OscarDHdz**
## Hey, I just made a Pull Request!
* Fixes [4209](https://github.com/backstage/backstage/issues/4209)

### Problem Statement
If your company has a wide range of templates, your backstage user would need to scroll through all templates looking for the one he needs. Or maybe, user just wants to know what type of components he can create using a specific technology. 😢 

### Solution 😎 
By adding Search and Filter functionalities, user now can easily and quickly find templates.

### Changes
* Using same Filter functionality as Catalog Page:
   * Updated logic to list available **categories** instead of **tags** (as it works in catalog page)
   * Updated Filter Groups:
      * Removed "Personal > Owned" 
      * Kept "My Company > All" and "Personal > Starred" group filters.  (_we can later add star feature for templates_)
   * I copied from `catalog` plugin:
      * `scaffolder>src>filter>*` - Renamed variables to align filtering behavior (Using categories instead of tags to filter)
      * `scaffolder>src>components>ResultsFilter>*` - Renamed variables to align filtering behavior
      * `scaffolder>src>components>ScaffolderFilter>*` - This is the same `CatalogFilter` from `catalog` plugin, it was just renamed to `ScaffolderFilter`
      * `scaffolder>src>hooks>*`
* Implemented Search functionality:
   * Added Search Input - Looks and feel the same way as it does in catalog page table 
   * Allows user to search templates by Entity `metadata` **tags** and **title** 
* Updated Layout to fit new controls as seen in screenshot:
   * Using same layout as Catalog Page, except that search input is located above the filter options so that the templates grid have more space to display cards. 

Screen Shot 2021-02-09 at 17 26 46

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx [hunk 1]
```diff
@@ -93,33 +145,59 @@ export const ScaffolderPage = () => {
...
+                <Link href="https://backstage.io/docs/features/software-templates/adding-templates">
+                  here!
+                </Link>
+              </Typography>
+            )}
+            {error && (
+              <WarningPanel>
+                Oops! Something went wrong loading the templates:{' '}
+                {error.message}
+              </WarningPanel>
```
Comment (adamdmharvey) [inline plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx:173]: 👍 for this!  I think ideally you provide the "Oops! Something went wrong loading the templates." as a `title={}` param, and leave the error.message for the child element inside the component. That way, the folded header says "Warning: Oops! Something went wrong loading the templates." and the user can click it to unfold to see the error?

File: plugins/scaffolder/src/hooks/useStarredEntities.test.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    mockStorage = new WebStorage('@backstage', new MockErrorApi()).forBucket(
+      Date.now().toString(), // TODO(blam): need something that changes every test run for now until the MockStorage is implemented
+    );
+  });
+  it('should return an empty set for when there is no items in storage', async () => {
+    const { result } = renderHook(() => useStarredEntities(), { wrapper });
+
+    expect(result.current.starredEntities.size).toBe(0);
+  });
+  it('should return a set with the current items when there is items in storage', async () => {
```
Comment (adamdmharvey) [inline plugins/scaffolder/src/hooks/useStarredEntities.test.tsx:68]:
```suggestion
  it('should return a set with the current items when there are items in storage', async () => {
```

File: plugins/scaffolder/src/hooks/useStarredEntities.test.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    // catch when the hook re-renders with the latest data
+    setTimeout(() => result.current.toggleStarredEntity(mockEntity), 1);
+
+    await waitForNextUpdate();
+
+    expect(result.current.starredEntities.size).toBe(1);
+    expect(result.current.isStarredEntity(mockEntity)).toBeTruthy();
+  });
+
+  it('should write new entries to the local store when adding a togglging entity', async () => {
```
Comment (adamdmharvey) [inline plugins/scaffolder/src/hooks/useStarredEntities.test.tsx:98]:
```suggestion
  it('should write new entries to the local store when adding a toggling entity', async () => {
```

**Review by adamdmharvey — CHANGES_REQUESTED** [review]
Wow great PR! 👏 well done.

Minor comments below, but an overall one: does this accidentally mix the concept of starring an ENTITY vs. starring a template?  i don't see where I'd be able as a user to star a template, like "Create React", which is different than my starring my resulting entity in the catalog called "MyReactWebSite"?  Apols I may not fully understand it so wanted to ask!

File: plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx (hunk 1 above)
Comment (OscarDHdz) [reply plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx:173]: Sure! Makes more sense. Updating it

File: plugins/scaffolder/src/hooks/useStarredEntities.test.tsx (hunk 3 above)
Comment (OscarDHdz) [reply plugins/scaffolder/src/hooks/useStarredEntities.test.tsx]: Fixed in 452602094f23e5bb0eb150e649c9661066d2f75e

File: plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx (hunk 1 above)
Comment (OscarDHdz) [reply plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx]: Fixed in 452602094f23e5bb0eb150e649c9661066d2f75e

**Comment by OscarDHdz** [comment]
> Wow great PR! 👏 well done.
> 
> Minor comments below, but an overall one: does this accidentally mix the concept of starring an ENTITY vs. starring a template? i don't see where I'd be able as a user to star a template, like "Create React", which is different than my starring my resulting entity in the catalog called "MyReactWebSite"? Apols I may not fully understand it so wanted to ask!

@adamdmharvey This is something my squad and I were wondering. We replicated the same filters from Catalog Page, we did remove the "Owned" filter, but regarding the "Star" we were not sure to remove it or not. We thought... "maybe later we can add a way to start templates" but that's something we wanted to discuss here with the Community 🤔   (_And yes, we thought about it in the sam as starring an ENTITY_). If there's no need/liking to have that "start template" feature, we can easily remove it from here 😎

**Comment by regicsolutions** [comment]
I think it would also be helpful to be able to filter by tags as well, thoughts?

**Comment by OscarDHdz** [comment]
> I think it would also be helpful to be able to filter by tags as well, thoughts?

Listing some pros/cons that I foresee:

**Filtering by Categories:**
Pros:
* Small/Restricted list of checkboxes - Only the defined type of components a user can create. 

Cons:
* No Tag filtering

**Filtering by Tags:**
Cons:
* Larger/(a bit)Unrestricted list of checkboxes - As they are added by people who create templates.
* No Category filter

**Both filers:**
Pros:
* All filter options!  

Cons:
* Larger list of checkboxes
* Might add confusion to the user if both elements are mixed

Notes:
* We could add another tab/panel/list to differentiate/split filters like one list for categories then on list for tags

**Comment by regicsolutions** [comment]
I see where you are coming from on keeping it clean and taking the less is more approach, and excellent job on this by the way!

My thought was the same tags may be used across templates and feel it enforces us to really only use tags that are meaningful... it would be beneficial for a developer only interested in templates that offer `logging-monitoring` to be able to filter by that also nice to be able to filter by tech-stack. How about if we add tags but keep it minimized by default?

Another option is the do what Spotify did and add it as part of search?

image

**Comment by OscarDHdz** [comment]
I like the idea of having collapse/expand controls to shrink the Filters pane. I think I'll give it a try 👍 
Also I'd like to see how the one from Spotify works 🤔 Probably we could work on that in a second iteration.

**Comment by benjdlambert** [comment]
So - i've been thinking about this PR on the side for a day or so now, and there's a few things that spring to mind. I don't intend to block merging of this PR based on any of the below, just some food for thought I think, and i'd be interested to get everyones opinion on it.

But first, thank you so much for doing this and the hard work you've put in. It looks great, and thanks for the great PR body explaining what you'd done to duplicate parts like filtering logic and hooks, its super helpful when reviewing and lets us know what we can maybe extract out into something separate. I appreciate that some of the code from the catalog table and filtering isn't so optimised for use across plugins. I know, I wrote most of it 🤷  👍 

So, To follow on from @regicsolutions point about how we do it internally at Spotify, things are a little different internally. These templates that you can select aren't part of a central catalog like they are in the Backstage Open Source. And the only way to see them is exactly that screenshot that @regicsolutions posted. The filtering at the top is purely text based too, it's client side I think, and just indexes the text that you can see on the React component, like owner, description, title, labels etc.

On our internal version, as you can see from the screenshot, they are ordered a little differently and grouped by their respective `type` I guess it would translate to in the OS world. That's something we don't do here, they're just coloured differently. Maybe grouping these templates in the frontend would remove the need for the category filter, as you can see it all grouped instead of it being unordered is a something that could help rather than category filters?

Also, for reference, internally we have around 30 templates, for different use cases and different reasons. Maybe what we have internally doesn't work for the open source side, as there's companies that have 100's of microtemplates and grouping them doesn't work at all 🤣   

I've also been thinking about the idea of just adding a `Templates` tab to the catalog view, and then just inheriting the filtering logic from there, but then it's a bit strange to be taken to the Entity page of the template rather than using that in the /create app to create a template from, and then you've got two places to see the same thing, but with different actions and different journeys, which might not make sense. But maybe there's something here that could be explored in the future too?

Like I said at the beginning of this wall of text, this is awesome, and it's a great feature. Just wanted to give you a little background as to what we do internally, and what i've been thinking. Maybe it doesn't make sense to duplicate the functionality from the catalog plugin at all later on down the line, but maybe our internal screenshot isn't the way to go either 👍

**Comment by mateiidavid** [comment]
Hey @benjdlambert had a read and I see a couple of points that I really like! 

To give a bit of context, @OscarDHdz has been working on this because at Expedia we plan on having quite a few templates, possibly totalling hundreds.

>  Maybe grouping these templates in the frontend would remove the need for the category filter, as you can see it all grouped instead of it being unordered and a purely text driven quick search is all you need?

This is super interesting, I think it would simplify the logic here drastically and have less of an overhead. My only question is on what basis would you group? Would it be on a specific tag? I'm also interested to see what @regicsolutions thinks here.

> I've also been thinking about the idea of just adding a Templates tab to the catalog view, and then just inheriting the filtering logic from there, but then it's a bit strange to be taken to the Entity page of the template rather than using that in the /create app to create a template from, and then you've got two places to see the same thing, but with different actions and different journeys, which might not make sense. But maybe there's something here that could be explored in the future too?

This was my first thought as well, especially since we are duplicating the logic quite a bit. We have a lot of services at Expedia though, I think creating from the entity view would be met with a bit of resistance by our adopters (even if I personally think it's a cleaner and more maintainable solution). From what I gather so far people have really grown accustomed to the different views. We could potentially add another action to it though? i.e just like services have `View/Edit` make it so that entities of kind template have a `create` action and re-direct to the form?

We can then leverage a more advanced search in the catalog (by filtering on tags, kinds, etc -- separate action can redirect to the create page) and keep a simpler text-based search and grouping for the actual plugin view.

We'd like to simplify this a bit if we can and offer a more consistent UX with the rest of the plugins. Oscar and I have been chatting and he has a few ideas. I was just spitballing here but yeah would love to hear what @regicsolutions has to say here since they have been involved in the whole process from issue creation to PR feedback :) 

@OscarDHdz

**Comment by regicsolutions** [comment]
Here's my view:

Been also thinking about this over the past few days... Here's where my head is at...

Taking a step back and simplifying things, Backstage is doing two things:

1. Acting as a **Service Registry** that keeps track of _existing_ components that have were built before backstage existed
2. Acting as a **Service Catalog** that acts as a developers gateway to select from a list of approved enterprise templates

I agree with @benjdlambert view as this should be visible in one view in a new tab by adding **Catalog** this is the same approach [kubeapps](https://www.youtube.com/watch?v=MHylHSDJqvc) is taking and i think it makes sense. so it would be: Services | Websites | Libraries | Documentation | Catalog | Other

I think the **create... +** link in the sidebar is redundant and there should only be two journey paths to getting to the Service Catalog:

1. Simply clicking on the Catalog tab for those users who know there way around backstage
2. For new comers, Click on on **+ Create Component** (adding a plus sign) where you are asked what you would like to do. @freben came up with this so i cant take any credit :), it was discussed as part of the recent catalog-import changes in PR and think it makes sense [4271](https://github.com/backstage/backstage/pull/4271)

106426127-57a77980-6465-11eb-8c0b-8889eae14724

Grouping the templates makes alot of sense and this will eliminate the need for "Categories" as they would be driven by groups and the granular filtering can be done using tags that are already in place, OpenShift does this today here's a view: 

catalog-browse-search-r3_6-1

image

When i change tabs the search thats already in place today should only search for that "type" so if i click on documentation it should only search for documentation

**Comment by freben** [comment]
Sweet, I love the discussions that got going in here!

@regicsolutions Careful with the wording here with "catalog". The catalog basically is a product in and of itself and its naming and use is deeply ingrained across Backstage. And the wording re "catalog" vs "service catalog" vs "software catalog" is a [discussion unto itself](https://github.com/backstage/backstage/issues/3065).

The way I look at the catalog - as in, the `/catalog` corner of things - is that it's technically inclined, management focused, dense "mechanical" tables, information rich, bird's eye view, special treatment of owned things - and should be modified to show ALL kinds - not just components - so its tabs at the top need to go away :)

Therefore, I do not at all feel worried about a duality between the catalog, and the `/create` corner of things. The latter is quite the opposite. It's fluffy, big friendly cards, show only the minimum of necessary information, no concept of ownership, specialised grouping and filtering - the user intent, visuals, mindset, and behaviour is _radically_ different. Mixing them together would probably do more harm than good :)

@benjdlambert What do you think about avoiding the "component" wording? I think it is pretty certain that components will be the most common thing to create, but there's bound to be more stuff that it'll be used for! Why not for example a "create managed storage" template that adorns an existing component repo with the config files necessary to spin up a managed database for it to use (with the corresponding Resource kind catalog info data)?

@mateiidavid It's fascinating to think of hundreds of individual templates! May I ask how it came to be that they are so many? One of the primary benefits of the software templates feature is to drive unification, reduce variation, push towards a smaller set of standards. Do absolutely not take this as bashing the idea, I just want to learn and make sure the feature works well for your case too. Maybe there's a way we could make the system let you design just ten templates, but with a dynamic "aspect"/"variant" that's one of the parameters to the templates? Feel free to throw me a DM if you don't want to talk too much about specifics in public.

**Comment by freben** [comment]
@adamdmharvey You had a blocker on this, good to go now from your point of view?

**Review by adamdmharvey — APPROVED** [review]
My comments addressed; approve to unlock my change request.  👍

File: plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx [hunk 4]
```diff
@@ -47,24 +61,62 @@ const getTemplateCardProps = (
...
+    ],
+    [isStarredEntity, orgName],
+  );
+  const [search, setSearch] = useState('');
+  const [matchingEntities, setMatchingEntities] = useState(
+    [] as TemplateEntityV1alpha1[],
   );
 
+  const matchesQuery = (metadata: EntityMeta, query: string) =>
+    `${metadata.title}`.toUpperCase().indexOf(query) !== -1 ||
```
Comment (freben) [inline plugins/scaffolder/src/components/ScaffolderPage/ScaffolderPage.tsx:107]: fyi there's also `'string'.includes('i') -> true`

File: plugins/scaffolder/src/hooks/useStarredEntities.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import { storageApiRef, useApi } from '@backstage/core';
+import { useCallback, useEffect, useState } from 'react';
+import { useObservable } from 'react-use';
+
+const buildEntityKey = (component: Entity) =>
+  `entity:${component.kind}:${component.metadata.namespace ?? 'default'}:${
+    component.metadata.name
+  }`;
+
+export const useStarredEntities = () => {
```
Comment (freben) [inline plugins/scaffolder/src/hooks/useStarredEntities.ts:27]: Oh, I really should move this to `@backstage/plugin-catalog-react` later for everybody to use.

**Review by freben — APPROVED** [review]
Hey - this is a really, really welcome addition, thanks. I had a few comments but do not want to block on them. I would appreciate it if you had the chance to look at them and address at least the two actionable ones (field filtering, and the `??` mistake) separately.

File: plugins/scaffolder/src/components/ScaffolderFilter/AllServicesCount.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import { useApi } from '@backstage/core';
+import { catalogApiRef } from '@backstage/plugin-catalog-react';
+import { CircularProgress, useTheme } from '@material-ui/core';
+import React from 'react';
+import { useAsync } from 'react-use';
+
+export const AllServicesCount = () => {
+  const theme = useTheme();
+  const catalogApi = useApi(catalogApiRef);
+  const { value, loading } = useAsync(() => catalogApi.getEntities());
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/scaffolder/src/components/ScaffolderFilter/AllServicesCount.tsx:26]: Come to think of it by the way - this one was actually meant to filter by kind Template as well, wasn't it? 🤔

**Comment by OscarDHdz** [comment]
@freben Sure! I'll take care of your comments today 😄

File: plugins/scaffolder/src/components/ScaffolderFilter/AllServicesCount.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder/src/components/ScaffolderFilter/AllServicesCount.tsx:26]: In the end, I didn't even need this `AllServiceCount` 😮 Got rid of it

File: plugins/scaffolder/src/hooks/useStarredEntities.ts (hunk 5 above)
Comment (OscarDHdz) [reply plugins/scaffolder/src/hooks/useStarredEntities.ts:27]: Moved and updated both `scaffolder` and `catalog` plugins

**Comment by dtuite** [comment]
Hi @OscarDHdz. I'm going to mention this PR in my [Backstage newsletter](https://roadie.io/backstage-weekly/) tomorrow. If you'd like the PR attributed to your employer or another entity please shoot me an email at david@roadie.io.

**Pull request merged**